% !TEX root = ../PhD Thesis.tex
\chapter{Objectives}

During my PhD, I developed transcriptomic web apps with graphical interfaces to be used by the scientific community, namely researchers with basic computational skills.

First, based on the work I developed during my MSc's thesis \cite{saraiva-agostinho:2016vw}, I continued working on psichomics \cite{saraiva-agostinho:2018uq,saraiva-agostinho:2020wz}, an alternative splicing quantification, analysis and visualisation R package for pre-processed human data from TCGA \cite{chang:2013ww}. During my PhD, psichomics was extended to support more data sources (including GTEx \cite{lonsdale:2013uo}, recount2 \cite{collado-torres:2017uw} and user-provided data), analyse gene expression and support alternative splicing quantification for 14 different species, among other features.

Using the Connectivity Map (CMap), a public database containing millions of gene expression changes in response to induced molecular and pharmacological perturbations \cite{subramanian:2017ul}, our lab also developed cTRAP, an R package to identify candidate causal perturbations from differential gene expression data, as well as predict compounds that may promote or revert them. cTRAP also allows to list putative targeting drugs based on drug sensitivity datasets and includes a GSEA-based enrichment analysis of molecular descriptors for compounds from NCI60 and CMap.

Both psichomics and cTRAP feature web-based graphical interfaces to assist users interactively performing most of their functions following easy steps, properly detailed in online tutorials that are constantly updated according to user feedback. To make the tools more accessible and freely available via any modern web browser, we also developed an app server to deploy psichomics, cTRAP and multiple other R packages as web apps -- including other programs built by my lab colleagues (voyAGEr, betAS and scStudio). I kindly invite you to pause, sit back and relax, visit our website at \alink{https://compbio.imm.medicina.ulisboa.pt}, wander through the web apps there and enjoy the journey. The landing page is a gallery of work from our lab that I am deeply proud to support.
