% !TEX root = ../PhD Thesis.tex
\chapter*{Summary}
\addcontentsline{toc}{chapter}{Summary}
% min. 300 words
% 5 keywords

During my PhD at Instituto de Medicina Molecular Jo√£o Lobo Antunes (iMM), I developed web apps for transcriptomic data analyses as free, open-source resources and an app server to deploy them.

\section*{psichomics}

Alternative pre-mRNA splicing generates functionally distinct transcripts from the same gene and is involved in the control of multiple cellular processes, with its dysregulation being linked to a variety of pathologies. The advent of next-generation sequencing has enabled global studies of alternative splicing in different physiologic and pathologic contexts. However, bioinformatics tools for alternative splicing analysis from RNA-seq data used to be user-unfriendly, disregard available exon-exon junction quantification or have limited downstream analysis features.

To overcome such limitations, we developed psichomics, an R package with an intuitive graphical interface for alternative splicing quantification and integrative analyses of alternative splicing and gene expression from large transcriptomic datasets, including those from The Cancer Genome Atlas (TCGA), the Genotype-Tissue Expression (GTEx) project, and the recount2 project, as well as user-provided data. psichomics assists the user in integrating sample-associated features (molecular and clinical) to perform survival, dimensionality reduction, and differential alternative splicing and gene expression analyses. Since its publication in 2018, psichomics has been used to discover splicing-associated prognostic factors and therapeutic targets, along with studying alternative splicing regulation in physiological and pathological contexts.

\section*{cTRAP}

The Connectivity Map (CMap) hosts differential expression profiles associated with thousands of genetic and pharmacologic perturbations (perturbagens) of human cells. We developed the cTRAP R package to identify potentially causal molecular perturbations by comparing user-provided differential gene expression results with those from CMap, using correlation and gene set enrichment scores. cTRAP can also compare against gene expression/drug sensitivity associations derived from the NCI-60 cancer cell line panel, the Cancer Therapeutics Response Portal and the Genomics of Drug Sensitivity in Cancer project, to pinpoint compounds that may target the phenotypes associated with the user-provided differential expression profiles. We envisage cTRAP allowing users to identify putative causal perturbations to better understand the molecular mechanisms associated with the observed phenotypes, as well as to predict therapeutic targets.

\section*{CompBio app server}

Both psichomics and cTRAP feature graphical interfaces to assist users in exploring most of their functionality. We set up the CompBio app server based on Docker Compose to deploy our lab's web apps, publicly available at \alink{compbio.imm.medicina.ulisboa.pt}.\\

\textbf{Keywords:} bioinformatics, web apps, alternative splicing, gene expression, perturbagens.

